common/management/src/role/role_mgr.rs

```diff
@@ -0,0 +1,199 @@
+            role_prefix: format!("{}/{}", ROLE_API_KEY_PREFIX, tenant),
+        }
+    }
+
+    async fn upsert_role_info(
+        &self,
+        role_info: &RoleInfo,
+        seq: Option<u64>,
+    ) -> common_exception::Result<u64> {
+        let key = format!("{}/{}", self.role_prefix, &role_info.name);
+        let value = serde_json::to_vec(&role_info)?;
+
+        let match_seq = match seq {
+            None => MatchSeq::GE(1),
+            Some(s) => MatchSeq::Exact(s),
+        };
+
+        let kv_api = self.kv_api.clone();
+        let upsert_kv = async move {
+            kv_api
+                .upsert_kv(UpsertKVAction::new(
+                    &key,
+                    match_seq,
+                    Operation::Update(value),
+                    None,
+                ))
+                .await
+        };
+        let res = upsert_kv.await?;
+        match res.result {
+            Some(SeqV { seq: s, .. }) => Ok(s),
+            None => Err(ErrorCode::UnknownRole(format!(
+                "unknown role, or seq not match {}",
+                role_info.name
+            ))),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RoleMgrApi for RoleMgr {
+    async fn add_role(&self, role_info: &RoleInfo) -> common_exception::Result<u64> {
+        let match_seq = MatchSeq::Exact(0);
+        let key = format!("{}/{}", self.role_prefix, &role_info.name);
+        let value = serde_json::to_vec(&role_info)?;
+
+        let kv_api = self.kv_api.clone();
+        let upsert_kv = kv_api.upsert_kv(UpsertKVAction::new(
+            &key,
+            match_seq,
+            Operation::Update(value),
+            None,
+        ));
+        let res = upsert_kv.await?.into_add_result()?;
+        match res.res {
+            OkOrExist::Ok(v) => Ok(v.seq),
+            OkOrExist::Exists(v) => Err(ErrorCode::UserAlreadyExists(format!(
+                "Role already exists, seq [{}]",
+                v.seq
+            ))),
+        }
+    }
+
+    async fn get_role(&self, role_name: &str, seq: Option<u64>) -> Result<SeqV<RoleInfo>> {
+        let key = format!("{}/{}", self.role_prefix, role_name);
+        let kv_api = self.kv_api.clone();
+        let get_kv = async move { kv_api.get_kv(&key).await };
+        let res = get_kv.await?;
+        let seq_value =
+            res.ok_or_else(|| ErrorCode::UnknownRole(format!("unknown role {}", role_name)))?;
+
+        match MatchSeq::from(seq).match_seq(&seq_value) {
+            Ok(_) => Ok(seq_value.into_seqv()?),
+            Err(_) => Err(ErrorCode::UnknownRole(format!(
+                "unknown role {}",
+                role_name
+            ))),
+        }
+    }
+
+    async fn get_roles(&self) -> Result<Vec<SeqV<RoleInfo>>> {
+        let role_prefix = self.role_prefix.clone();
+        let kv_api = self.kv_api.clone();
+        let prefix_list_kv = async move { kv_api.prefix_list_kv(role_prefix.as_str()).await };
+        let values = prefix_list_kv.await?;
```

Comment:
Why put it into an `async move` block?

`let values = kv_api.prefix_list_kv(role_prefix.as_str()).await?`
Is this just enough?

Reply:
i also have this question, it seems all the methods in this mod are in this style...  https://github.com/datafuselabs/databend/blob/9ad3602fb5905a170feec018bc1ba013eee7b935/common/management/src/user/user_mgr.rs#L135

is there any consideration about this? 😳

Reply:
If it compiles without the `async move` block, there won't be a problem IMO... 🤔

Reply:
thank you, lemme fix it